idunn/api/urlsolver.py

```diff
@@ -0,0 +1,34 @@
+from hashlib import sha1
+from secrets import token_bytes
+from urllib.parse import quote
+
+import httpx
+from fastapi import HTTPException
+from fastapi.responses import RedirectResponse
+
+
+client = httpx.AsyncClient()
+
+# Randomized seed used to hash addresses to solve, do not leek.
+random_token = token_bytes(64)
+
+
+def hash_url(url: str) -> str:
+    """
+    Hash of the URL that the client must provide in order to avoid abusive use
+    of the endpoint.
+    """
+    return sha1(random_token + url.encode()).hexdigest()
```

Comment:
sha1 is no longer considered safe for cryptographic use, so I would use sha256 instead.

Also, Python provides [`hmac.HMAC` ](https://docs.python.org/3.8/library/hmac.html#hmac.HMAC.hexdigest) to achieve this in a normalized way:
```suggestion
    return hmac.HMAC(key=random_token, msg=url.encode(), digestmod="sha256").hexdigest()
```

Reply:
I feel like a dinosaur ^^'

The doc also recommends to use [`compare_digest`](https://docs.python.org/3.8/library/hmac.html#hmac.compare_digest) to check provided hash. I don't see any reasonable reason why someone would make the effort of putting together a timing attack just to open a single URL, but given the tiny cost I suppose that we're better safe than sorry :D